src/utils/display-observer.js

```diff
@@ -0,0 +1,226 @@
+/**
+ * Copyright 2020 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {getServiceForDoc, registerServiceBuilderForDoc} from '../service';
+import {pushIfNotExist, removeItem} from './array';
+import {rethrowAsync} from '../log';
+
+const SERVICE_ID = 'DisplayObserver';
+
+/** @implements {Disposable} */
+export class DisplayObserver {
+  /**
+   * @param {!AmpDoc} ampdoc
+   */
+  constructor(ampdoc) {
+    const {win} = ampdoc;
+
+    /** @private @const {!IntersectionObserver} */
+    this.docObserver_ = new win.IntersectionObserver(
+      (e) => this.observed_(e, 0),
+      {
+        root: ampdoc.getBody(),
+        threshold: 0.001,
```

Comment:
what does this threshold do?

Reply:
I brought this up while ago. The way the InOb spec is defined, if something is just off-the-screen (e.g. `transform: translateX(100vw)`, InOb still considers it intersecting (`isIntersecting == true`) because it touches the edge of the viewport by 1px. You can't see it - but it's still intersecting. See this [demo](https://jsbin.com/juhoram/quiet). It's possible of course to do things like `translateX(calc(100vw + 1px))`, but it's guaranteed that no one would do that. The `threshold: 0.001` is the safest way I found to definitively say: this content is offscreen.

Reply:
Another good example is a scroller. E.g. something like this:

```
<div class="scroller">
  <div class="item"></div>
  <div class="item"></div>
</div>

.scroller {overflow: auto; display: flex;}
.item {flex-basis: 100%}
```

The second `item` would be logically and practically out-of-viewport, but the `isIntersecting == true`, b/c InOb would still think that they overlap by 1px.

Reply:
Ah! Right! You did mention that once. Clever fix. Thank you for the thorough explanation